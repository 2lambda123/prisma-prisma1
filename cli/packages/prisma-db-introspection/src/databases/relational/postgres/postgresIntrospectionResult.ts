import { ITable, ITableRelation } from "../relationalConnector"
import { RelationalIntrospectionResult } from "../relationalIntrospectionResult"
import { ISDL, IGQLField, IGQLType, IDirectiveInfo, plural, camelCase, capitalize, DatabaseType, Renderer, TypeIdentifier, TypeIdentifiers } from 'prisma-datamodel'
import * as _ from 'lodash'

export class PostgresIntrospectionResult extends RelationalIntrospectionResult {
  constructor(model: ITable[], relations: ITableRelation[], renderer?: Renderer) {
    super(model, relations, DatabaseType.postgres, renderer)
  }

  protected isTypeReserved(type: IGQLType): boolean {
    return type.name == '_RelayId'
  }
  protected toTypeIdentifyer(typeName: string): TypeIdentifier | null {
    switch(typeName) {
      case 'int1':
      case 'int2':
      case 'int4': 
      case 'int8': return TypeIdentifiers.integer
      case 'numeric':
      case 'float4':
      case 'float8': return TypeIdentifiers.float
      case 'varchar':
      case 'bpchar':
      case '_text':
      case 'text': return TypeIdentifiers.string
      case 'bool': return TypeIdentifiers.boolean
      case 'jsonb':
      case 'json': return TypeIdentifiers.json
      case '_date':
      case 'date':
      case 'timestamp': return TypeIdentifiers.dateTime
      case 'uuid': return TypeIdentifiers.uuid
      default: return null
    }
  }
  protected parseDefaultValue(defaultValueString: string, type: string): string | null {
    let val = defaultValueString

    // Detect string
    if(val.startsWith('\'')) {
      // Strip quotes (are added again by renderer)
      val = val.substring(1, val.length - 1)
    }

    // Remove cast operator
    const i = val.indexOf('::')

    if(i >= 0) {
      val = val.substring(0, i)
    }

    // Check for null
    if(val.toUpperCase() === 'NULL') {
      return null
    }
    
    // If the field is not a string field, 
    // and the default val is not a boolean or a number, we assume a function call or sequence reference. 
    if(type !== TypeIdentifiers.string) {
      if(isNaN(val as any) && val.toLowerCase() !== 'true' && val.toLowerCase() !== 'false') {
        return null
      }
    }

    // TODO: Sequences are simply ignored.

    return val
  }

}
