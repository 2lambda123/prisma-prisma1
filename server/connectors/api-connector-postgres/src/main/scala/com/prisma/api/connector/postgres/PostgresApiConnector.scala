package com.prisma.api.connector.postgres

import com.prisma.api.connector.ApiConnector
import com.prisma.api.connector.jdbc.impl.{JdbcDataResolver, JdbcDatabaseMutactionExecutor}
import com.prisma.config.DatabaseConfig
import com.prisma.shared.models.{ConnectorCapabilities, ConnectorCapability, Project, ProjectIdEncoder}

import scala.concurrent.{ExecutionContext, Future}

case class PostgresApiConnector(config: DatabaseConfig, isActive: Boolean, isPrototype: Boolean)(implicit ec: ExecutionContext) extends ApiConnector {
  lazy val databases = PostgresDatabasesFactory.initialize(config)

  override def initialize() = {
    databases
    Future.unit
  }

  override def shutdown() = {
    for {
      _ <- databases.primary.database.shutdown
      _ <- databases.replica.database.shutdown
    } yield ()
  }

  override val databaseMutactionExecutor = {
    val manageRelayIds = if (isPrototype) false else isActive
    JdbcDatabaseMutactionExecutor(databases.primary, manageRelayIds)
  }
  override def dataResolver(project: Project)       = JdbcDataResolver(project, databases.primary)
  override def masterDataResolver(project: Project) = JdbcDataResolver(project, databases.primary)
  override def projectIdEncoder: ProjectIdEncoder   = ProjectIdEncoder('$')
  override val capabilities = {
    if (isPrototype) {
      ConnectorCapabilities.postgresPrototype
    } else {
      ConnectorCapabilities.postgres(isActive = isActive)
    }
  }
}
