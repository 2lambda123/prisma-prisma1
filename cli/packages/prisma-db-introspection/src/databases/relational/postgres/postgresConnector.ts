import {
  RelationalConnector,
  ITable,
  IColumn,
  ITableRelation,
  IIndex,
} from '../relationalConnector'
import * as _ from 'lodash'
import { Client } from 'pg'
import { TypeIdentifier, DatabaseType } from 'prisma-datamodel'
import { PostgresIntrospectionResult } from './postgresIntrospectionResult'
import { RelationalIntrospectionResult } from '../relationalIntrospectionResult'
import { PrismaDBClient } from '../../prisma/prismaDBClient'

// Documentation: https://www.prisma.io/docs/data-model-and-migrations/introspection-mapping-to-existing-db-soi1/

// Responsible for extracting a normalized representation of a PostgreSQL database (schema)
export class PostgresConnector extends RelationalConnector {
  client: Client

  constructor(client: Client) {
    super()

    if (
      !(client instanceof Client) &&
      !((client as any) instanceof PrismaDBClient)
    ) {
      throw new Error('Postgres instance needed for initialization.')
    }

    this.client = client
  }

  public getDatabaseType(): DatabaseType {
    return DatabaseType.postgres
  }

  protected createIntrospectionResult(
    models: ITable[],
    relations: ITableRelation[],
  ): RelationalIntrospectionResult {
    return new PostgresIntrospectionResult(models, relations)
  }

  protected async query(query: string, params?: any[]): Promise<any[]> {
    return (await this.client.query(query, params)).rows
  }

  public async listSchemas(): Promise<string[]> {
    const schemas = await super.listSchemas()
    return schemas.filter(schema => !schema.startsWith('pg_'))
  }

  // TODO: Unit test for column comments
  protected async queryColumnComment(
    schemaName: string,
    tableName: string,
    columnName: string,
  ) {
    const commentQuery = `
      SELECT
      (
        SELECT
          pg_catalog.col_description(c.oid, cols.ordinal_position::int)
        FROM pg_catalog.pg_class c
        WHERE
          c.oid     = (SELECT ('"' || cols.table_schema || '"."' || cols.table_name || '"')::regclass::oid) AND
          c.relname = cols.table_name
      ) as column_comment   
      FROM
        information_schema.columns cols
      WHERE
        cols.table_schema  = $1::text AND
        cols.table_name    = $2::text AND
        cols.column_name   = $3::text;
    `
    const [comment] = (await this.query(commentQuery, [
      schemaName,
      tableName,
      columnName,
    ])).map(row => row.column_comment as string)

    if (comment === undefined) {
      return null
    } else {
      return comment
    }
  }

  protected async queryIndices(schemaName: string, tableName: string) {
    const indexQuery = `
      SELECT
          tableInfos.relname as table_name,
          indexInfos.relname as index_name,
          array_to_string(array_agg(columnInfos.attname), ',') as column_names,
          rawIndex.indisunique as is_unique,
          rawIndex.indisprimary as is_primary_key
      FROM
          -- pg_class stores infos about tables, indices etc: https://www.postgresql.org/docs/9.3/catalog-pg-class.html
          pg_class tableInfos,
          pg_class indexInfos,
          -- pg_index stores indices: https://www.postgresql.org/docs/9.3/catalog-pg-index.html
          pg_index rawIndex,
          -- pg_attribute stores infos about columns: https://www.postgresql.org/docs/9.3/catalog-pg-attribute.html
          pg_attribute columnInfos,
          -- pg_namespace stores info about the schema
          pg_namespace schemaInfo
      WHERE
          -- find table info for index
          tableInfos.oid = rawIndex.indrelid
          -- find index info
          AND indexInfos.oid = rawIndex.indexrelid
          -- find table columns
          AND columnInfos.attrelid = tableInfos.oid
          AND columnInfos.attnum = ANY(rawIndex.indkey)
          -- we only consider oridnary tables
          AND tableInfos.relkind = 'r'
          -- we only consider stuff out of one specific schema
          AND tableInfos.relnamespace = schemaInfo.oid
          AND schemaInfo.nspname = $1::text
          AND tableInfos.relname = $2::text
      GROUP BY
          tableInfos.relname,
          indexInfos.relname,
          rawIndex.indisunique,
          rawIndex.indisprimary
    `
    return (await this.query(indexQuery, [schemaName, tableName])).map(row => {
      return {
        tableName: row.table_name as string,
        name: row.index_name as string,
        fields: this.parseIndexColumns(row.column_names),
        unique: row.is_unique as boolean,
        isPrimaryKey: row.is_primary_key as boolean,
      }
    })
  }

  private parseIndexColumns(arrayAsString: string): string[] {
    return arrayAsString.split(',').map(x => x.trim())
  }
}
