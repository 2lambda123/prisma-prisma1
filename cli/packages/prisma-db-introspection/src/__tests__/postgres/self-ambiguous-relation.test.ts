import Connectors from '../../connectors'
import { Client } from 'pg'
import { connectionDetails } from './connectionDetails'
import { PostgresConnector } from '../../databases/relational/postgres/postgresConnector'
import { DatabaseType } from 'prisma-datamodel'
import { connect } from 'tls'

async function introspect(client: Client) {
  return (await Connectors.create(DatabaseType.postgres, client).introspect('databaseintrospector')).renderToNormalizedDatamodelString()
}

async function testSchema(sql: string) {
  const client = new Client(connectionDetails)

  await client.connect()
  await client.query('DROP SCHEMA IF EXISTS DatabaseIntrospector cascade;')
  await client.query('CREATE SCHEMA DatabaseIntrospector;')
  await client.query('SET search_path TO DatabaseIntrospector;')
  await client.query(sql)

  expect(await introspect(client)).toMatchSnapshot()

  await client.end()
}

describe('Introspector', () => {
  // Schema from https://github.com/prismagraphql/prisma/issues/2504
  test('test schema - 2504', async () => {
    await testSchema(sql)
  })
})

const sql = `
/*******************************************************************************
 Create Tables
********************************************************************************/

CREATE TABLE "Employee"
(
  "EmployeeId" INT NOT NULL,
  "LastName" VARCHAR(20) NOT NULL,
  "FirstName" VARCHAR(20) NOT NULL,
  "Title" VARCHAR(30),
  "ReportsTo" INT,
  "BirthDate" TIMESTAMP,
  "HireDate" TIMESTAMP,
  "Address" VARCHAR(70),
  "City" VARCHAR(40),
  "State" VARCHAR(40),
  "Country" VARCHAR(40),
  "PostalCode" VARCHAR(10),
  "Phone" VARCHAR(24),
  "Fax" VARCHAR(24),
  "Email" VARCHAR(60),
  CONSTRAINT "PK_Employee" PRIMARY KEY  ("EmployeeId")
);

/*******************************************************************************
 Create Primary Key Unique Indexes
********************************************************************************/

/*******************************************************************************
 Create Foreign Keys
********************************************************************************/

ALTER TABLE "Employee" ADD CONSTRAINT "FK_EmployeeReportsTo"
  FOREIGN KEY ("ReportsTo") REFERENCES "Employee" ("EmployeeId") ON DELETE NO ACTION ON UPDATE NO ACTION;

CREATE INDEX "IFK_EmployeeReportsTo" ON "Employee" ("ReportsTo");
`
