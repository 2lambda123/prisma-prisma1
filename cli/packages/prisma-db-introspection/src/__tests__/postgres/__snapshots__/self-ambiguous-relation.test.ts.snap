// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Introspector test schema - 2504 1`] = `
"type Employee {
  Address: String
  BirthDate: DateTime
  City: String
  Country: String
  Email: String
  Fax: String
  FirstName: String!
  HireDate: DateTime
  id: Int! @pgColumn(name: \\"EmployeeId\\")
  LastName: String!
  Phone: String
  PostalCode: String
  ReportsTo: Employee
  State: String
  Title: String
}"
`;
